qiskit/circuit/qpy_serialization.py

```diff
@@ -0,0 +1,712 @@
+                                          qbit.register.name.encode('utf8'))
+        file_obj.write(instruction_arg_raw)
+    for clbit in instruction_tuple[2]:
+        instruction_arg_raw = struct.pack(INSTRUCTION_ARG_PACK,
+                                          'c'.encode('utf8'),
+                                          clbit.index,
+                                          clbit.register.name.encode('utf8'))
+        file_obj.write(instruction_arg_raw)
+    # Encode instruction params
+    for param in instruction_tuple[0].params:
+        container = io.BytesIO()
+        if isinstance(param, int):
+            type_key = 'i'
+            data = struct.pack('<q', param)
+            size = struct.calcsize('<q')
+        elif isinstance(param, float):
+            type_key = 'f'
+            data = struct.pack('<d', param)
+            size = struct.calcsize('<d')
+        elif isinstance(param, Parameter):
+            type_key = 'p'
+            data = struct.pack(PARAMETER_PACK, param._name.encode('utf8'),
+                               param._uuid.bytes)
+            size = PARAMETER_SIZE
+        elif isinstance(param, (np.integer, np.floating, np.ndarray)):
+            type_key = 'n'
+            np.save(container, param)
+            container.seek(0)
+            data = container.read()
+            size = len(data)
+        else:
+            raise TypeError("Invalid parameter type %s for gate %s," % (
+                instruction_tuple[0], type(param)))
+        instruction_param_raw = struct.pack(INSTRUCTION_PARAM_PACK,
+                                            type_key.encode('utf8'), size)
+        file_obj.write(instruction_param_raw)
+        file_obj.write(data)
+        container.close()
+
+
+def _write_custom_instruction(file_obj, name, instruction):
+    if isinstance(instruction, Gate):
+        type_str = 'g'.encode('utf8')
+    else:
+        type_str = 'i'.encode('utf8')
+    has_definition = False
+    size = 0
+    data = None
+    num_qubits = instruction.num_qubits
+    num_clbits = instruction.num_clbits
+    if instruction.definition:
+        has_definition = True
+        definition_buffer = io.BytesIO()
+        _write_circuit(definition_buffer, instruction.definition)
+        definition_buffer.seek(0)
+        data = definition_buffer.read()
+        definition_buffer.close()
+        size = len(data)
+    custom_instruction_raw = struct.pack(CUSTOM_DEFINITION_PACK,
+                                         name.encode('utf8'), type_str,
+                                         num_qubits, num_clbits,
+                                         has_definition, size)
+    file_obj.write(custom_instruction_raw)
+    if data:
+        file_obj.write(data)
+
+
+def dump(file_obj, circuits):
+    """Write QPY binary data to a file
+
+    This function is used to save a circuit to a file for later use or transfer
+    between machines. The output QPY file is forward compatible and can be loaded
```

Comment:
My understanding is that qpy offers *backwards* compatibility rather than forwards compatibility. Ie, newer qiskit can load qpy generated by older qiskit.

Reply:
I was probably wording this awkwardly, I was thinking from the perspective of the output file, it's forward compatible in the same way a CD is forwards compatible with a DVD player. qpy as a spec/format in general is backwards compatible. But I agree that's confusing. I'll adjust this now.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5578/commits/e810ca0ec172041043547c47baf464e8d33f5695